### PR TITLE
Encryption: support upload local encrypted files in DMFile to S3

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8351

Problem Summary:

### What is changed and how it works?

- Files in S3 are always not encrypted by TiFlash.
- Implement `StdStreamFromReadBuffer` which inherits `std::iostream` and can be passed to `Aws::S3::Model::PutObjectRequest::SetBody`
- Read local file -> decrypt -> upload to S3

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Remote/DataStore/DataStoreS3.cpp [hunk 1]
```diff
@@ -103,6 +133,7 @@ bool DataStoreS3::putCheckpointFiles(
 
     std::vector<std::future<void>> upload_results;
     // upload in parallel
+    // Note: Local checkpoint files are always not encrypted.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Remote/DataStore/DataStoreS3.cpp:136]: We will not write checkpoint files in the local disk after https://github.com/pingcap/tiflash/pull/8385 merged.

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Comment by Lloyd-Pottiger** [comment]
> Since we have inevitably written DMFiles to the local disks, can we avoid using `S3WritableFile` to reduce the number of calls to S3. For example, we can implement a 'DecryptStream' to replace `Aws::FStream` : https://github.com/pingcap/tiflash/blob/master/dbms/src/Storages/S3/S3Common.cpp#L599-L600.

may not be a good idea, since functions in `Aws::FStream` like `open`, `close` are not virtual functions.

**Comment by Lloyd-Pottiger** [comment]
> > Since we have inevitably written DMFiles to the local disks, can we avoid using `S3WritableFile` to reduce the number of calls to S3. For example, we can implement a 'DecryptStream' to replace `Aws::FStream` : https://github.com/pingcap/tiflash/blob/master/dbms/src/Storages/S3/S3Common.cpp#L599-L600.
> 
> may not be a good idea, since functions in `Aws::FStream` like `open`, `close` are not virtual functions.

o, maybe we can inherit `basic_iostream`.

File: dbms/src/Encryption/FileProvider.h [hunk 2]
```diff
@@ -56,6 +56,7 @@ class FileProvider
         const WriteLimiterPtr & write_limiter_ = nullptr,
         int flags = -1,
         mode_t mode = 0666) const;
+    static WritableFilePtr newS3WritableFile(const String & file_key);
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/FileProvider.h]: yeah, removed

File: dbms/src/IO/StdStreamBufFromReadBuffer.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+namespace DB
+{
+class ReadBuffer;
+class ReadBufferFromFileBase;
+
+/// `std::streambuf`-compatible wrapper around a ReadBuffer.
+/// Warning:
+/// The implementation this class is quite tricky and not complete.
+/// We need to continuously keep the pointers of get area in sync by calling `setg()` and changing `position()`.
+/// Please use it with caution and do enough tests.
```
Comment (JaySon-Huang) [inline dbms/src/IO/StdStreamBufFromReadBuffer.h:30]: Can we add some ut about the known issue of `StdStreamBufFromReadBuffer` and `StdStreamFromReadBuffer`?

File: dbms/src/Server/Server.cpp [hunk 4]
```diff
@@ -1134,6 +1134,7 @@ int Server::main(const std::vector<std::string> & /*args*/)
 
     global_context->getSharedContextDisagg()->initRemoteDataStore(
         global_context->getFileProvider(),
+        global_context->getReadLimiter(),
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Server/Server.cpp]: ok, remove RateLimiter

File: dbms/src/IO/StdStreamBufFromReadBuffer.h (hunk 3 above)
Comment (Lloyd-Pottiger) [reply dbms/src/IO/StdStreamBufFromReadBuffer.h:30]: currently no known issue, I have added some ut about common usage of reading from std::stream;

File: dbms/src/Interpreters/SharedContexts/Disagg.h [hunk 5]
```diff
@@ -16,6 +16,7 @@
 
 #include <Core/TiFlashDisaggregatedMode.h>
 #include <Encryption/FileProvider_fwd.h>
+#include <Encryption/RateLimiter.h>
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Interpreters/SharedContexts/Disagg.h]: removed

File: dbms/src/Storages/S3/S3Common.cpp [hunk 6]
```diff
@@ -596,10 +604,20 @@ static bool doUploadFile(
...
+    {
+        local_file = file_provider->newRandomAccessFile(local_fname, encryption_path, nullptr);
+    }
+    else
+    {
+        local_file = PosixRandomAccessFile::create(local_fname);
+    }
+    // Read at most 16MB each time.
+    auto read_buf = std::make_unique<ReadBufferFromRandomAccessFile>(local_file, 16 * 1024 * 1024);
+    auto istr = std::make_shared<StdStreamFromReadBuffer>(std::move(read_buf), write_bytes);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/S3Common.cpp:619]:
```suggestion
    auto istr = Aws::MakeShared<StdStreamFromReadBuffer>("PutObjectInputStream", std::move(read_buf), write_bytes);
```

Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3Common.cpp:619]: The `istr` finally will be released in `PutObjectRequest`. Use `Aws::MakeShared` for it or ASAN may throw warning message like https://github.com/pingcap/tiflash/pull/8618

**Review by JaySon-Huang — COMMENTED** [review]
Rest LGTM

File: dbms/src/Storages/S3/S3Common.cpp [hunk 7]
```diff
@@ -596,10 +604,20 @@ static bool doUploadFile(
...
+    if (file_provider)
+    {
+        local_file = file_provider->newRandomAccessFile(local_fname, encryption_path, nullptr);
+    }
+    else
+    {
+        local_file = PosixRandomAccessFile::create(local_fname);
+    }
+    // Read at most 16MB each time.
+    auto read_buf = std::make_unique<ReadBufferFromRandomAccessFile>(local_file, 16 * 1024 * 1024);
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3Common.cpp:618]: And if the `write_bytes` is smaller, use the `write_bytes` as buffer size

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/S3Common.cpp:618]: > Can we make 16MB smaller? In theory, to speed up the upload process, we can use higher concurrency to upload files.

ok, change to `min(8MB, write_bytes)`, the same as `dt_segment_delta_small_column_file_size`.

File: dbms/src/Storages/S3/S3Common.cpp (hunk 6 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/S3Common.cpp]: done

File: dbms/src/Storages/S3/S3Common.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/S3Common.cpp]: In theory, no.

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/S3Common.cpp]: change to 1MB.

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/S3Common.cpp]: The default buffer of `std::fstream` is 1KB which maybe too small, and we should align to the OS page size 4KB

**Comment by JaySon-Huang** [comment]
```
[2024-01-12T09:55:37.009Z] FAILED TESTS (1/2444):
[2024-01-12T09:55:37.009Z]    96449 ms: /tiflash/gtests_dbms ComputeServerRunner.multipleQuery
```
https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-unit-tests/detail/tiflash-ghpr-unit-tests/12553/pipeline

**Pull request merged**